### PR TITLE
iOS 10309 crouton on screen

### PR DESCRIPTION
**Pull request by WanaldinoTelefonica**
## 🎟️ **Jira ticket**
[IOS-10309](https://jira.tid.es/browse/IOS-10309)

## 🥅 **What's the goal?**
- Allow crouton to be shown in a specific viewController

## 🚧 **How do we do it?**
- Store the viewController on the ongoingCroutons (weak reference as we don't want to retain it)
- If the viewController is defined, don't look for the top most viewController

### CODE REVIEW

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift [hunk 1]
```diff
@@ -33,20 +39,20 @@ public extension CroutonController {
...
+//        style: CroutonStyle = .info,
+//        dismissHandler: DismissHandlerBlock? = nil
+//    ) -> Token {
+//        showCrouton(
+//            config: config,
+//            style: style,
+//            dismissHandler: dismissHandler,
+//            rootViewController: UIApplication.shared.windows.filter(\.isKeyWindow).first?.rootViewController
+//        )
+//    }
```
Comment (WanaldinoTelefonica) [inline Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift:56]: This is the old version without the deploy parameter. To remove

File: Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+//
+//            if let rootVC = visibleVC as? UINavigationController {
+//                return visibleViewController(from: rootVC.topViewController!)
+//            } else if let homeVC = visibleVC.parent?.tabBarController,
+//                      let selectedNavigationController = homeVC.selectedViewController as? UINavigationController {
+//                return visibleViewController(from: selectedNavigationController.topViewController!)
+//            } else {
+//                return visibleVC
+//            }
+//        }
```
Comment (WanaldinoTelefonica) [inline Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift:66]: This method was on the other class to find the viewController. I think this is not usefull anymore. To be removed

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift [hunk 3]
```diff
@@ -9,9 +9,16 @@
 import UIKit
 
 public class CroutonController: NSObject {
+    public enum RootViewController {
+        public typealias Closure = () -> UIViewController?
+        public static let `default`: Closure = { UIApplication.shared.windows.filter(\.isKeyWindow).first?.rootViewController }
+    }
+
+//    public typealias RootViewControllerClosure = () -> UIViewController?
```
Comment (idenjoe) [inline Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift:17]: Remove commented code

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift [hunk 4]
```diff
@@ -9,9 +9,16 @@
...
+//    public static let defaultRootViewControllerClosure = { UIApplication.shared.windows.filter(\.isKeyWindow).first?.rootViewController }
+
```
Comment (idenjoe) [inline Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift:19]: Remove commented code

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift (hunk 1 above)
Comment (idenjoe) [reply Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift:56]: Do not integrate this without removing the commented code.

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift (hunk 4 above)
Comment (dhidalgofadrique) [reply Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift:19]: +1. If not used, better removing it (could be restored from history if needed)

File: Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift (hunk 2 above)
Comment (dhidalgofadrique) [reply Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift:66]: If not used, remove it

File: Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            }
+        }
+
+        private func viewController() -> UIViewController? {
+            if let exactViewController {
+                return exactViewController
+            } else if let rootViewController = rootViewController() {
+                return visibleViewController(from: rootViewController)
+            } else {
+                return nil
```
Comment (dhidalgofadrique) [inline Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift:46]: Has it sense not having viewController? If not, maybe an `assertionFailure` could be useful here

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift (hunk 1 above)
Comment (WanaldinoTelefonica) [reply Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift]: Removed

File: Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift (hunk 5 above)
Comment (WanaldinoTelefonica) [reply Sources/Mistica/Components/Crouton/Presentation/OngoingCrouton.swift:47]: Done

File: Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift (hunk 4 above)
Comment (WanaldinoTelefonica) [reply Sources/Mistica/Components/Crouton/Presentation/CroutonController.swift]: Removed

**Review by dhidalgofadrique — APPROVED** [review]
Good job

**Review by idenjoe — APPROVED** [review]

**Pull request merged**